design-proposals/klip-3-serialization-of-single-fields.md

```diff
@@ -0,0 +1,1072 @@
+selection of primitive types (`Number`, `String`, and `Boolean`); Avro supports
+array and map types along with a wider range of primitive types.
+
+The collective term _record_ is used to refer to JSON _objects_ and Avro _records_. 
+The term _non-record_ is used to refer to any other types.
+
+A row in KSQL has a logical schema made up of multiple fields. Fields within the row's
+schema are persisted to either the key or value of the underlying Kafka message. 
+In this sense the key and value also have a logical schema.  
+
+The term _single field_ is used to describe a key or value logical schema where 
+the schema contains only a single field. 
+           
+## Motivation and background
+
+Currently, the `JSON` and `AVRO` formats in KSQL can only deserialize _record_ values, 
+i.e. those in the form of a JSON object or an Avro record. These formats can 
+only deserialize _non-record_ types if the type is contained as a field within a
+top-level _record_ type.
+
+Likewise, KSQL always serializes the value fields with in a schema as fields within
+a top-level _record_. There is no support for serializing schemas that contain only 
+a single value field as _non-record_ types. For example, if a query selects only a 
+single `INT` value field, it will be serialized within a _record_, it can not be 
+serialized as a simple integer.
+
+On the other hand, KSQL only supports string keys, i.e. a _non-record_ type.
+
+Being able to both process and produce both _record_ and _non-record_ values opens up 
+KSQL to more use cases, and is a precondition of enhancing KSQL to support _record_
+keys, a.k.a. structured keys. (Support for _non_record_ keys will be needed to 
+allow KSQL to maintain backwards compatibility with existing queries, which have
+string keys).
+
+For example, given a statement such as: 
+
+```
+CREATE STREAM USERS (NAME STRING) WITH (...);
+ ``` 
+ 
+The logical value schema has only a single field. KSQL should be able to read 
+data in Kafka where the value is a anonymous top-level JSON or Avro string, rather 
+than only a _record_ containing a `NAME` field.
+
+Given a statement such as:
+
+```
+CREATE STREAM FOO AS SELECT ID FROM BAR;
+ ``` 
+ 
+The _logical_ schema is again only a single field, lets say of type `INT`. 
+Users should be able to control if this is serialized as a named `ID` field within
+_record_, or as an anonymous JSON number or Avro int. 
+ 
+
+## What is in scope
+
+* JSON (de)serialization of schemas containing a single value field that 
+  is an `ARRAY`, `MAP`, `STRUCT` or primitive type.
+  
+* Avro (de)serialization of schemas containing a single value field that
+  is an `ARRAY`, `MAP`, `STRUCT` or primitive type.  
+  
+* (De)serialization of _record_ and _non-record_ keys, but only in the context on
+  ensuring a design that can be extended to include such keys.
+  
+* Any configuration and/or syntax used to control the (de)serialization of single fields.
+
```

Comment:
possibly make clear here, using hte same vocabulary, what from the existing behavior will be retained ? (i think all of it, but helps to have a clear picture, in one place, of how the whole complex area will work once this is implemented)

Reply:
I'm not sure I follow what you're getting at here.  Are you referring to maintaining backwards compatibility?